File: pwa/src/host-version.test.ts
import { describe, expect, it } from 'vitest'
import { clientIsAhead, compareVersions, hostKindNoun, hostVersionLine } from './host-version'

/**
 * The comparator behind the one sentence this client says about the gap between
 * its build and the host's — *update this server from a desktop* — and the label
 * beside the version number. Both are display; neither acts.
 */

describe('compareVersions', () => {
  it('orders release segments left to right, missing treated as zero', () => {
    expect(compareVersions('0.10.0', '0.9.9')).toBe(1)
    expect(compareVersions('0.9.9', '0.10.0')).toBe(-1)
    expect(compareVersions('1.0.0', '1.0.0')).toBe(0)
    expect(compareVersions('1.2', '1.2.0')).toBe(0)
    expect(compareVersions('2', '1.9.9')).toBe(1)
  })

  it('sorts a prerelease below the release it belongs to', () => {
    expect(compareVersions('1.0.0-rc.1', '1.0.0')).toBe(-1)
    expect(compareVersions('1.0.0', '1.0.0-rc.1')).toBe(1)
    expect(compareVersions('1.0.0-rc.2', '1.0.0-rc.1')).toBe(1)
  })

  it('ignores a leading v and build metadata', () => {
    expect(compareVersions('v1.2.3', '1.2.3')).toBe(0)
    expect(compareVersions('1.2.3+build.9', '1.2.3+build.1')).toBe(0)
  })
})

describe('clientIsAhead — the whole of when the update sentence shows', () => {
  it('is true only when the client build is strictly greater', () => {
    expect(clientIsAhead('0.10.0', '0.9.0')).toBe(true)
    expect(clientIsAhead('1.0.0', '1.0.0-rc.1')).toBe(true)
  })

  it('is false when the client is equal or behind', () => {
    expect(clientIsAhead('0.10.0', '0.10.0')).toBe(false)
    expect(clientIsAhead('0.9.0', '0.10.0')).toBe(false)
  })

  it('is false — never a guess — when either number is not a real version', () => {
    // '' is what an older host sends and what this client holds before a socket
    // is up; 'unknown' is version.ts's honest non-answer. Comparing against
    // either would manufacture a verdict, so the answer is no and the sentence
    // stays off the screen.
    expect(clientIsAhead('', '0.9.0')).toBe(false)
    expect(clientIsAhead('0.10.0', '')).toBe(false)
    expect(clientIsAhead('unknown', '0.9.0')).toBe(false)
    expect(clientIsAhead('0.10.0', 'unknown')).toBe(false)
    expect(clientIsAhead('unknown', 'unknown')).toBe(false)
  })
})

describe('hostKindNoun', () => {
  it('calls a headless host a server and a desktop a desktop', () => {
    expect(hostKindNoun('headless')).toBe('server')
    expect(hostKindNoun('desktop')).toBe('desktop')
  })

  it('gives no noun for a kind that was never said', () => {
    expect(hostKindNoun(null)).toBeNull()
  })
})

describe('hostVersionLine', () => {
  it('names the version and the kind when both are known', () => {
    expect(hostVersionLine('0.10.0', 'headless')).toBe('version 0.10.0 · server')
    expect(hostVersionLine('1.2.3', 'desktop')).toBe('version 1.2.3 · desktop')
  })

  it('names only the version when the kind was never said', () => {
    expect(hostVersionLine('0.10.0', null)).toBe('version 0.10.0')
  })

  it('is empty for a host that reported no version, so the caller draws nothing', () => {
    expect(hostVersionLine('', 'headless')).toBe('')
    expect(hostVersionLine('', null)).toBe('')
  })
})

File: android/app/src/test/java/dev/terminaldeck/android/protocol/HostVersionTest.kt
package dev.terminaldeck.android.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The version arithmetic the "update this server from a desktop" sentence hangs on.
 *
 * Mirrors `pwa/src/host-version.ts`; the risk is not that a comparison can be written but that this
 * copy and the browser's disagree about whether a phone is ahead — which would show the nudge on one
 * client and not the other for the same pair of builds.
 */
class HostVersionTest {

    @Test
    fun `release segments compare left to right with a missing one as zero`() {
        assertTrue(HostVersion.compare("0.10.0", "0.9.9") > 0)
        assertTrue(HostVersion.compare("0.9.0", "0.10.0") < 0)
        assertEquals(0, HostVersion.compare("1.2.3", "1.2.3"))
        assertEquals(0, HostVersion.compare("1.2", "1.2.0"))
        assertTrue(HostVersion.compare("v0.10.0", "0.10.0") == 0)
    }

    @Test
    fun `a prerelease sorts below the release it belongs to`() {
        assertTrue(HostVersion.compare("1.0.0-beta", "1.0.0") < 0)
        assertTrue(HostVersion.compare("1.0.0", "1.0.0-beta") > 0)
        assertTrue(HostVersion.compare("1.0.0-alpha", "1.0.0-beta") < 0)
        assertTrue(HostVersion.compare("1.0.0-alpha.1", "1.0.0-alpha.2") < 0)
        // Numeric identifiers rank below alphanumeric ones.
        assertTrue(HostVersion.compare("1.0.0-1", "1.0.0-alpha") < 0)
        // Build metadata takes no part in precedence.
        assertEquals(0, HostVersion.compare("1.0.0+abc", "1.0.0+def"))
    }

    @Test
    fun `ahead is default-closed unless both numbers are real and the client is strictly greater`() {
        assertTrue(HostVersion.clientIsAhead("0.10.0", "0.9.0"))
        assertFalse(HostVersion.clientIsAhead("0.9.0", "0.10.0"))
        assertFalse(HostVersion.clientIsAhead("0.10.0", "0.10.0"))
        // A non-answer on either side manufactures no verdict.
        assertFalse(HostVersion.clientIsAhead("0.10.0", ""))
        assertFalse(HostVersion.clientIsAhead("", "0.9.0"))
        assertFalse(HostVersion.clientIsAhead("unknown", "0.9.0"))
        assertFalse(HostVersion.clientIsAhead("0.10.0", "unknown"))
    }

    @Test
    fun `the kind noun is server for headless and desktop for desktop and nothing for the rest`() {
        assertEquals("server", HostVersion.hostKindNoun("headless"))
        assertEquals("desktop", HostVersion.hostKindNoun("desktop"))
        assertNull(HostVersion.hostKindNoun(null))
        assertNull(HostVersion.hostKindNoun("something-else"))
    }

    @Test
    fun `the version line names the kind when there is one and draws nothing without a version`() {
        assertEquals("version 0.10.0 · server", HostVersion.hostVersionLine("0.10.0", "headless"))
        assertEquals("version 0.10.0 · desktop", HostVersion.hostVersionLine("0.10.0", "desktop"))
        assertEquals("version 0.10.0", HostVersion.hostVersionLine("0.10.0", null))
        assertEquals("", HostVersion.hostVersionLine("", "headless"))
    }

    @Test
    fun `the behind sentence appears only when ahead and names the right kind of box`() {
        val server = HostVersion.behindSentence("0.10.0", "0.9.0", "headless")
        assertTrue(server != null && server.contains("server"))
        val desktop = HostVersion.behindSentence("0.10.0", "0.9.0", "desktop")
        assertTrue(desktop != null && desktop.contains("desktop"))
        // Not ahead: nothing to say.
        assertNull(HostVersion.behindSentence("0.9.0", "0.10.0", "headless"))
        assertNull(HostVersion.behindSentence("0.10.0", "", "headless"))
    }
}
